test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeModule.cs

```diff
@@ -130,31 +130,71 @@ public Task UpdateDesiredPropertiesAsync(object patch, CancellationToken token)
-                        },
-                        null,
-                        TimeSpan.FromSeconds(5),
-                        token);
-                },
+                () => this.WaitForReportedPropertyUpdatesInternalAsync(expected, token),
                 "Received expected twin updates for module '{Module}'",
                 this.Id);
         }
+
+        protected Task WaitForReportedPropertyUpdatesInternalAsync(object expected, CancellationToken token)
+        {
+            return Retry.Do(
+                async () =>
+                {
+                    Twin twin = await this.iotHub.GetTwinAsync(this.deviceId, this.Id, token);
+                    return twin.Properties.Reported;
+                },
+                reported => JsonEquals((expected, "properties.reported"), (reported, string.Empty)),
+                null,
+                TimeSpan.FromSeconds(5),
+                token);
+        }
+
+        // reference.rootPath and comparand.rootPath are path strings like those returned from
+        // Newtonsoft.Json.Linq.JToken.Path, and compatible with the path argument to
+        // Newtonsoft.Json.Linq.JToken.SelectToken(path)
+        static bool JsonEquals(
+            (object obj, string rootPath) reference,
+            (object obj, string rootPath) comparand)
+        {
+            Dictionary<string, JValue> ProcessJson(object obj, string rootPath)
+            {
+                // return all json values under root path, with their relative
+                // paths as keys
+                int length = rootPath.Length + (rootPath == string.Empty ? 0 : 1);
+                return JObject
+                    .FromObject(obj)
+                    .SelectToken(rootPath)
+                    .Cast<JContainer>()
+                    .DescendantsAndSelf()
+                    .OfType<JValue>()
+                    .ToDictionary(v => v.Path.Substring(length));
+            }
+
+            var referenceValues = ProcessJson(reference.obj, reference.rootPath);
+            var comparandValues = ProcessJson(comparand.obj, comparand.rootPath);
+
+            // comparand equals reference if, for each json value in reference:
+            // - comparand has a json value with the same path
+            // - the json values match
+            bool match = referenceValues.All(kvp => comparandValues.ContainsKey(kvp.Key) &&
+                kvp.Value.Equals(comparandValues[kvp.Key]));
+
+            if (!match)
+            {
+                var missing = referenceValues
+                    .Where(kvp =>
+                        !comparandValues.ContainsKey(kvp.Key) ||
+                        !kvp.Value.Equals(comparandValues[kvp.Key]))
+                    .Select(kvp => kvp.Key);
+                Log.Verbose(
+                    "Expected configuration values missing in agent's reported properties:\n  {MissingValues}",
```

Comment:
Not just missing, but not matching also, right? If so I would recommend changing the var name above as well.

Reply:
I was thinking about it in the strictest sense: If there's a similar-but-not-exactly-the-same property in the object (e.g., the name is the same but the value is off by one character), then that's no more helpful than if the property wasn't there at all. The expected value is present (exactly as specified) or it's missing.